dotnet/src/Skills/Skills.MsGraph/TaskListSkill.cs

```diff
@@ -37,19 +37,19 @@ public static class Parameters
     }
 
     private readonly ITaskManagementConnector _connector;
-    private readonly ILogger<TaskListSkill> _logger;
+    private readonly ILogger _logger;
 
     /// <summary>
     /// Initializes a new instance of the <see cref="TaskListSkill"/> class.
     /// </summary>
     /// <param name="connector">Task list connector.</param>
-    /// <param name="logger">Logger.</param>
-    public TaskListSkill(ITaskManagementConnector connector, ILogger<TaskListSkill>? logger = null)
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    public TaskListSkill(ITaskManagementConnector connector, ILoggerFactory? loggerFactory = null)
     {
         Ensure.NotNull(connector, nameof(connector));
 
         this._connector = connector;
-        this._logger = logger ?? new NullLogger<TaskListSkill>();
+        this._logger = loggerFactory is not null ? loggerFactory.CreateLogger(nameof(TaskListSkill)) : NullLogger.Instance;
```

Comment:
What are your thoughts about the generic versions of ILogger<T> and ILoggerFactory.CreateLogger<T>()?

i.e. syntax sugaring this line:
`var logger = kernel.LoggerFactory?.CreateLogger<TaskListSkill>() ?? NullLogger<TaskListSkill>.Instance;`

Reply:
I think there are actually three questions here:
1. Should this use `ILogger<T>` instead of `ILogger`?
2. Should this use `NullLogger<T>.Instance` instead of `NullLogger.Instance`?
3. Should this use `CreateLogger<T>()` instead of `CreateLogger(nameof(T))`?

For (1), `ILogger<T>` is a strange API. Its sole purpose to my knowledge is in support of DI, so that you can have an `ILogger<Foo>` in your ctor and DI will end up supplying it with the result of effectively `GetService<ILoggerFactory>().CreateLogger<Foo>())`, such that it's an ILogger created with the "Foo" category; the actual `ILogger<T>` interface isn't any more useful to a consumer than just `ILogger`. So, in this use (and every other similar use in SK), there's no need for `_logger` to be an `ILogger<TaskListSkill>`, just `ILogger`.

For (2), a NullLogger doesn't do any logging, so category doesn't matter. And since with (1) we just decided we don't need `ILogger<T>`, there's no benefit to using `NullLogger<T>.Instance` instead of `NullLogger.Instance`.

(3) is the interesting one. The purpose of all of `CreateLogger<T>()`, `CreateLogger(typeof(T))`, and `CreateLogger(string)` is to supply a category to the logger; that's it, nothing more. The latter is the cheapest: it's calling to the actual interface method on ILoggerFactory, just providing the category name.  The built in LoggerFactory will use that string to look up an `ILogger` instance in its cache, and just return the existing one it has: it tries to only ever create one `ILogger` instance per category and return the same instance for all requests for that category. `CreateLogger(typeof(T))` is an extension method that creates a string from the specified Type; in .NET 8's nuget package for this, that's been made much cheaper, at least, although there's still a bit of logic executed to determine that it can just use Type.FullName (when it can... it can't in some situations)... on earlier nuget packages, this always ends up creating a new string.  And then there's `CreateLogger<T>()`. This one is frustrating, as it _always_ ends up allocating a new `Logger<T>` instance: that's because it's statically typed to return an `ILogger<T>` rather than an `ILogger`, but ILoggerFactory.CreateLogger returns an `ILogger`, and so it can't just delegate to the instance method as does `CreateLogger(Type)`.  All of that is to say, the cheapest answer here is to use `CreateLogger(nameof(Whatever))`, and switching to `CreateLogger<Whatever>()` adds unnecessary and unavoidable cost.  In a case like this, where it's done once per skill, that overhead doesn't really matter; it's more impactful in situations where you might be getting a logger per log entry, in which case you really want to use a cached instance over and over again.  Hence why I went with always using `CreateLogger(string)`.  I could switch the ones that aren't hot path to be `CreateLogger<T>` if you'd prefer.

Reply:
fwiw I've rewritten the logger approach in a separate repo, using ILogger<T> and replacing NullLogger<T> with DefaultLogger<T>. The final result has improved a lot:
* DefaultLogger follows settings in appsettings.json and appsettings.development.json
* ILogger<T> makes logs more readable thanks to the inclusion of the class name, and more configurable with filters in appsettings.json

Reply:
```csharp
public static class DefaultLogger<T>
{
    public static readonly ILogger<T> Instance = GetLogFactory()
        .CreateLogger<T>();

    private static ILoggerFactory GetLogFactory()
    {
        return WebApplication.CreateBuilder().Build().Services.GetService<ILoggerFactory>()
               ?? throw new ConfigurationException("Unable to provide logger factory");
    }
}
```

example usage in a ctor:

```csharp
public AzureCognitiveSearchMemory(
    AzureCognitiveSearchConfig config,
    ILogger<AzureCognitiveSearchMemory>? log = null)
{
    this._log = log ?? DefaultLogger<AzureCognitiveSearchMemory>.Instance;
```